backend/src/schema/resolvers/emails.js

```diff
@@ -0,0 +1,92 @@
+import generateNonce from './helpers/generateNonce'
+import Resolver from './helpers/Resolver'
+import existingEmailAddress from './helpers/existingEmailAddress'
+import { UserInputError } from 'apollo-server'
+import Validator from 'neode/build/Services/Validator.js'
+
+export default {
+  Mutation: {
+    AddEmailAddress: async (_parent, args, context, _resolveInfo) => {
+      let response
+      try {
+        const { neode } = context
+        await new Validator(neode, neode.model('UnverifiedEmailAddress'), args)
+      } catch (e) {
+        throw new UserInputError('must be a valid email')
+      }
+
+      // check email does not belong to anybody
+      await existingEmailAddress(_parent, args, context)
+
+      const nonce = generateNonce()
+      const {
+        user: { id: userId },
+      } = context
+      const { email } = args
+      const session = context.driver.session()
+      const writeTxResultPromise = session.writeTransaction(async txc => {
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-02T08:59:52Z" title="Wednesday, October 2nd 2019, 10:59:52 am +02:00">Oct 2, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -0,0 +1,92 @@
+import generateNonce from './helpers/generateNonce'
+import Resolver from './helpers/Resolver'
+import existingEmailAddress from './helpers/existingEmailAddress'
+import { UserInputError } from 'apollo-server'
+import Validator from 'neode/build/Services/Validator.js'
+
+export default {
+  Mutation: {
+    AddEmailAddress: async (_parent, args, context, _resolveInfo) => {
+      let response
+      try {
+        const { neode } = context
+        await new Validator(neode, neode.model('UnverifiedEmailAddress'), args)
+      } catch (e) {
+        throw new UserInputError('must be a valid email')
+      }
+
+      // check email does not belong to anybody
+      await existingEmailAddress(_parent, args, context)
+
+      const nonce = generateNonce()
+      const {
+        user: { id: userId },
+      } = context
+      const { email } = args
+      const session = context.driver.session()
+      const writeTxResultPromise = session.writeTransaction(async txc => {
```


fancy stuff here! this is the first time in the code base we are leveraging this, I believe..

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-02T09:07:53Z" title="Wednesday, October 2nd 2019, 11:07:53 am +02:00">Oct 2, 2019</time>_

---


could you summarize the benefits of using `writeTransaction` and `writeTxResultPromise`?
or point me to some docs that explain it?

Actually, I found this `// It is possible to execute write transactions that will benefit from automatic retries
// on both single instance ('bolt' URI scheme) and Causal Cluster ('bolt+routing' URI scheme)` from https://neo4j.com/docs/api/javascript-driver/current/
Are automatic retries the main reason you decided to use this here?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-10-02T13:22:43Z" title="Wednesday, October 2nd 2019, 3:22:43 pm +02:00">Oct 2, 2019</time>_

---


Yes, I have no idea how to check those benefits. E.g. how to simulate them. But the docs clearly say: Do this. So I did.

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-10-02T14:31:55Z" title="Wednesday, October 2nd 2019, 4:31:55 pm +02:00">Oct 2, 2019</time>_

---


it seems like it would be a benefit to have automatic retries, and I'm in favor of following best practices. :)